examples/ReAct/hotpotqa/inference.py

```diff
@@ -0,0 +1,164 @@
+        return HotpotqaState(step_idx=1, last_state="", current_state=current_state, buffered_action="")
+
+    def step(self, state, action):
+        """Take a step in the world model.
+        
+        :param state: the current state (see the docstring of BlocksWorldModel)
+        :param action: the action to take
+        :return: the next state and additional information cached for reward calculation
+        """
+
+        state = copy.deepcopy(state)
+        buffered_action = state.buffered_action
+        current_state = state.current_state
+        step_idx = state.step_idx
+        current_state = self.update_state(current_state, action, step_idx)
+        new_buffered_action = action
+
+        if action == "exceed maxlength":
+            self.terminal = True
+
+        if self.terminal == True:
+            state = HotpotqaState(step_idx=self.max_steps, last_state=None,
+                        current_state=None, buffered_action= "Finish[]")
+            aux = {"ans_finished": self.terminal}
+            return state, aux
+
+        state = HotpotqaState(step_idx=step_idx+1, last_state=state.current_state,
+                        current_state=current_state, buffered_action=new_buffered_action)
+        aux = {"ans_finished": self.terminal}
+
+        return state, aux
+
+    def update_state(self, current_state: str, action: HotpotqaAction, step_idx) -> str:
+        """Update the block states with the action.
+
+        :param block_states: the current block states. Note that this argument is a string,
+            and it's only a part of 'BWState'
+        :param action: the action to take
+        :return: the updated block states
+        """
+        try:
+            thought, action = action.strip().split(f"\nAction {step_idx}: ")
+        except:
+            thought = action.strip().split('\n')[0]
+            if step_idx != 1 and thought == "":
+                self.terminal = True
+                return None
+            new_state = current_state + f" {thought}\nAction {step_idx}:"
+        
+        new_state = current_state + utils.webthink(self.base_model, step_idx, action, thought)
+
+        return new_state
+
+    def is_terminal(self, state: HotpotqaState) -> bool:
+        if self.terminal == True:
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        if utils.finished_check(state.buffered_action):
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        if state.step_idx == self.max_steps:
+            self.base_model.reset()
+            self.terminal = False
+            return True
+        return False
+
+def main(model_dir, llama_size="8B", prompt="examples/ReAct/hotpotqa/prompts/react.json", data_path="examples/ReAct/hotpotqa/data/hotpot_dev_v1_simplified.json"):
+    
+    base_model = Llama3Model(model_dir, llama_size, max_batch_size=1, max_seq_len=20000)
+    
+    env = wikienv.WikiEnv()
```

Comment:
Let's create a tool module, so that user can call the tool like:
```
from reasoners.tools import wikipedia, calculator

toolset = [wikipedia, calculator]

world_model = ReactWorldModel(toolset)
```

Just an example. Should look at other frameworks (e.g., AutoAgent, Langchain, BMTools) to define the interface